graylog2-server/src/main/java/org/graylog/plugins/views/migrations/V20250224150000_SearchJobStateCleanupOnStartup.java

```diff
@@ -0,0 +1,75 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * Server Side Public License for more details.
+ *
+ * You should have received a copy of the Server Side Public License
+ * along with this program. If not, see
+ * <http://www.mongodb.com/licensing/server-side-public-license>.
+ */
+package org.graylog.plugins.views.migrations;
+
+import com.google.common.collect.Sets;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.model.Filters;
+import jakarta.inject.Inject;
+import org.graylog.plugins.views.search.db.SearchJobService;
+import org.graylog.plugins.views.search.errors.SimpleSearchError;
+import org.graylog.plugins.views.search.jobs.SearchJobState;
+import org.graylog.plugins.views.search.jobs.SearchJobStateService;
+import org.graylog.plugins.views.search.jobs.SearchJobStatus;
+import org.graylog2.database.MongoCollections;
+import org.graylog2.migrations.Migration;
+import org.graylog2.plugin.system.NodeId;
+
+import java.time.ZonedDateTime;
+import java.util.Set;
+
+import static org.graylog.plugins.views.search.jobs.SearchJobState.STATUS_FIELD;
+
+/**
+ * This migration runs at every startup to clean SearchJobs up that were running while Graylog stopped/restarted
+ * and moves them into an error state so that they can be restarted manually by the user.
+ */
+public class V20250224150000_SearchJobStateCleanupOnStartup extends Migration {
+
+    private final MongoCollection<SearchJobState> collection;
+    private final SearchJobStateService searchJobStateService;
+    private final SearchJobService searchJobService;
+    private final NodeId nodeId;
+
+    @Inject
+    public V20250224150000_SearchJobStateCleanupOnStartup(final MongoCollections mongoCollections,
+                                                          final SearchJobStateService searchJobStateService,
+                                                          final SearchJobService searchJobService,
+                                                          final NodeId nodeId) {
+        this.collection = mongoCollections.collection(SearchJobStateService.COLLECTION_NAME, SearchJobState.class);
+        this.searchJobStateService = searchJobStateService;
+        this.searchJobService = searchJobService;
+        this.nodeId = nodeId;
+    }
+
+    @Override
+    public ZonedDateTime createdAt() {
+        return ZonedDateTime.parse("2025-02-24T15:00:00Z");
+    }
+
+    @Override
+    public void upgrade() {
+        // find all jobs in state "RUNNING" on startup
+        collection.find(Filters.eq(STATUS_FIELD, SearchJobStatus.RUNNING)).forEach(job -> {
+            // if this job is supposed to run on this node and it is not in the cache, it can't be a running job and should be moved to error state
+            if(job.identifier().executingNodeId().equals(nodeId.getNodeId()) && (searchJobService.getFromCache(job.id()) == null)) {
```

Comment:
Do migrations run only on the leader?

If yes, this code is probably wrong, because you will inject leader's `searchJobService`.
You may find a RUNNING job from a different node, realize it is not in the cache of leader's `searchJobService` (but it may be in a cache of another node's `searchJobService`).

If no, why not include current node id in MongoDB query?

Reply:
Imho they run only on the leader, from `ServerBootstrap`:

        try {
            if (**configuration.isLeader()** && configuration.runMigrations()) {
                runMigrations(injector, MigrationType.STANDARD);
            }
        } catch (Exception e) {
            LOG.error("Exception while running migrations", e);
            System.exit(1);
        }

Reply:
If my concerns are ok, maybe Periodical with properly set `runsForever` would be better? So that each node checks their own jobs once, on startup?

Reply:
thx. I addressed all your suggestions